plugins/async-node/core/src/index.test.ts

```diff
@@ -1,42 +1,147 @@
-        },
-      ],
-    },
+          id: 'uhh',
+          async: 'true'
+        }
+      ]
+    }
   ],
   navigation: {
-    BEGIN: "FLOW_1",
+    BEGIN: 'FLOW_1',
     FLOW_1: {
-      startState: "VIEW_1",
+      startState: 'VIEW_1',
       VIEW_1: {
-        state_type: "VIEW",
-        ref: "my-view",
-        transitions: {},
-      },
-    },
-  },
+        state_type: 'VIEW',
+        ref: 'my-view',
+        transitions: {}
+      }
+    }
+  }
+};
+
+const asyncNodeTest = async (resolvedValue: any, expectedActionType: string) => {
+  const plugin = new AsyncNodePlugin({
+    plugins: [new AsyncNodePluginPlugin()]
+  });
+
+  let deferredResolve: ((value: any) => void) | undefined;
+
+  let updateOnAsyncCounter = 0;
+
+  let resolverInstance: Resolver;
+  let beforeResolveCalled = false;
+
+  plugin.hooks.onAsyncNode.tap('test', async (node: Node.Node) => {
+    updateOnAsyncCounter++; // The Async Node can be tapped multiple times
+    return new Promise((resolve) => {
+      deferredResolve = resolve; // Promise would be resolved only once
+    });
+  });
+
+  let updateNumber = 0;
+
+  const player = new Player({ plugins: [plugin] });
+
+  player.hooks.viewController.tap('async-node-test', (vc) => {
+    vc.hooks.view.tap('async-node-test', (view) => {
+      view.hooks.onUpdate.tap('async-node-test', () => {
+        updateNumber++;
+      });
+    });
+  });
+
+  player.hooks.viewController.tap("async-node-test", (vc) => {
+    vc.hooks.view.tap("async-node-test", (view) => {
+      view.hooks.resolver.tap("async-node-test", (resolver) => {
+        resolverInstance = resolver;
+        resolverInstance.hooks.beforeResolve.tap("async-node-test", (node, options) => {
+          beforeResolveCalled = true;
+          return node; // return the original node if there's no meaningful Node to return
+        });
+      });
+    });
+  });
+
+  player.start(basicFRFWithActions as any);
+
+  let view = (player.getState() as InProgressState).controllers.view.currentView
+    ?.lastUpdate;
+
+  expect(view).toBeDefined();
+  expect(view?.actions[0].asset.type).toBe('action');
+  expect(view?.actions[1]).toBeUndefined();
+
+  await waitFor(() => {
+    expect(deferredResolve).toBeDefined();
+  });
+
+  // Consumer responds with null/undefined
+  if (deferredResolve) {
+    deferredResolve(resolvedValue);
+  }
+
+  await waitFor(() => {
+    expect(updateNumber).toBe(2);
+  });
+
+  view = (player.getState() as InProgressState).controllers.view.currentView
+    ?.lastUpdate;
+
+  expect(view?.actions[0].asset.type).toBe('action');
+  expect(updateNumber).toBe(2); // Replace with the actual expected number of updates
+
+  await waitFor(() => {
+    expect(resolverInstance).toBeDefined();
+    expect(beforeResolveCalled).toBe(true);
+  });
+
+  // Now we will force update the view using the resolver instance
+  const thirdForcedUpdate = resolverInstance.update();
+
+  if (deferredResolve) {
+    deferredResolve(thirdForcedUpdate);
```

Comment:
I'm not sure this will actually update anything, nothing is waiting for this to resolve, right?

Reply:
removed